### PR TITLE
Add start of dimensionality notebook doc

### PR DESCRIPTION
**Pull request by canyon289**
Adding start of shapes doc. Note for this PR I am not intending to reach "feature exhaustive" but more minimally useful that can be built upon.

There's two primary areas I could use with in this PR in ranked order

- [x] Feedback on the notebook to ensure the documentation is correct and useful
- [x] Help getting it to actually show up in the rendered html. The things Ive tried so far haven't seemed to work

Huge thanks to @michaelosthege who spent 2 hours in his Friday evening explaining this to me very patiently and kindly

### CODE REVIEW

File: docs/source/learn/examples/dimensionality.ipynb
Comment (canyon289) [inline docs/source/learn/examples/dimensionality.ipynb:1]: <p>Ill come back and fix this TODO in ~24 hours</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5437/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='0'/>

Comment (ricardoV94) [inline docs/source/learn/examples/dimensionality.ipynb:1]: <p>Perhaps, use shape if you would like to be "explicit" instead of "strict"?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5437/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='0'/>

Comment (ricardoV94) [inline docs/source/learn/examples/dimensionality.ipynb:1]: <p>Huh, shouldn't this  have been a vector of 2 elements?</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5437/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='25'/>

Comment (ricardoV94) [inline docs/source/learn/examples/dimensionality.ipynb:1]: <p>Reiterate this is the "implied" situation again?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5437/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='27'/>

Comment (michaelosthege) [reply docs/source/learn/examples/dimensionality.ipynb:1]: <p>Restart &amp; run all? If this is confirmed that'd be <em>very</em> concerning.</p>

Comment (michaelosthege) [reply docs/source/learn/examples/dimensionality.ipynb:1]: <p>A general advice regarding shape, probably worth adding at the top: __Pick prime numbers for dimension lenghts.__</p><p>Then one can always reconstruct how, for example, two variables were concatenated.</p><p>Also makes it easier to pinpoint errors.</p><p> </p><p>In this example, using <code>eye(3)</code>  would make it clear which output dimensions correspond to the support dim.</p><p> </p>

File: docs/source/learn/examples.md
Comment (OriolAbril) [inline docs/source/learn/examples.md]: the file is called `dimensionality`, it is not being included in the docs correctly with this.

**Review by OriolAbril — CHANGES_REQUESTED** [review]

**Comment by OriolAbril** [comment]
also, more of a PSA, I am not subscribed to notifications for pymc repo, only for pymc-examples. If there are docs PR please tag me so I see them asap. I try to review new issues and PRs 1-2 a week and I saw this one relatively early, but it is often too late already

File: docs/source/learn/examples/dimensionality.ipynb
Comment (OriolAbril) [reply docs/source/learn/examples/dimensionality.ipynb:1]: <p>it is indeed a vector of length 2: https://pymc--5437.org.readthedocs.build/en/5437/learn/examples/dimensionality.html#vector-distributions (notebooks in pymc are executed automatically every time the documentation is built, that includes PR previews)</p>

Comment (OriolAbril) [inline docs/source/learn/examples/dimensionality.ipynb:1]: <p>I would use a different description for <code>dims</code>.  They are not really a named array (I think, not sure what that is) but a list/array of dimension names.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5437/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='0'/>

Comment (OriolAbril) [inline docs/source/learn/examples/dimensionality.ipynb:1]: <p>use <code>{class}'arviz.InferenceData'</code>  (changing apostrophes for backticks) so that it is rendered as a link to arviz docs.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5437/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='20'/>

Comment (OriolAbril) [inline docs/source/learn/examples/dimensionality.ipynb:1]: <p>I think those should either be accompanied of a quick example (i.e. prior sampling plus inspecting the inferencedata result) or adding links to some relevant pages (i.e. the prior/posterior predictive one)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5437/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='30'/>

Comment (canyon289) [reply docs/source/learn/examples/dimensionality.ipynb:1]: <p>Reran, its length 2, sorry for the alarm</p>

Comment (canyon289) [reply docs/source/learn/examples/dimensionality.ipynb:1]: <p>Agreed. Im not going to do it in this PR. I'll open an issue and open it up to others or get back to it later so we dont lose track</p>

Comment (canyon289) [reply docs/source/learn/examples/dimensionality.ipynb:1]: <p>I dont quite get what you mean here, though I took a guess and added a new cell to show the limit of where Im losing the understanding</p>

**Comment by canyon289** [comment]
> also, more of a PSA, I am not subscribed to notifications for pymc repo, only for pymc-examples. If there are docs PR please tag me so I see them asap. I try to review new issues and PRs 1-2 a week and I saw this one relatively early, but it is often too late already

Noted @OriolAbril, sorry about that and thanks for reviewing

File: docs/source/learn/examples/dimensionality.ipynb
Comment (ricardoV94) [reply docs/source/learn/examples/dimensionality.ipynb:1]: <p>What Michael meant is that instead of having a (2, 2) distribution (repeated dimensions), you should use something like (2, 3), so that it's obvious what is base and what is implied/replicated dimensionality.</p><p> </p><p>Your example now needs a size 3 vectors for mu though. That's why it's failing</p>

File: docs/source/learn/examples/dimensionality.ipynb
Comment (ricardoV94) [inline docs/source/learn/examples/dimensionality.ipynb:1]: <p><code>mu=[0, 0]</code></p><p> </p><p>Technically, it's incorrect to specify a scalar mu for the multivariate normal, although for backwards compatibility we reshape mu behind the scenes</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc/pull/5437/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='25'/>

Comment (canyon289) [reply docs/source/learn/examples/dimensionality.ipynb:1]: If its just for backwards compability why dont we drop it?

Comment (michaelosthege) [reply docs/source/learn/examples/dimensionality.ipynb:1]: We can drop things after warning about them for long enough. If you want to drop supporting the automatic broadcast, then we'll need FutureWarning in a release version for about a year. After that we can make the break and argue that it's okay to not bump the major version.

Comment (twiecki) [reply docs/source/learn/examples/dimensionality.ipynb:1]: But we have a major release coming up where breaks are expected.

Comment (ricardoV94) [reply docs/source/learn/examples/dimensionality.ipynb:1]: We can drop it with a useful ValueError about the changed behavior. See also https://github.com/pymc-devs/pymc/issues/5440

Comment (canyon289) [reply docs/source/learn/examples/dimensionality.ipynb:1]: I agree with Thomas, we're releasing a major version this is *our chance* to reform the code without the need for FutureWarnings.

This is good for both devs and the community. For example in this case some of these incorrect but sort of works situations makes using the library more challenging. As a dev when I try and figure out what's happening I have to navigate the maze of conditionals and reference calls in the backend to figure out what is happening. I also need to read through so many much more code and tests to make a useful contribution.

In the nicest way I'm finding you (Ricardo and Michael) are PyMC/Aesara/AePPL power users and you're not inside the maze of code. You guys are so smart you have this elevated view and can see this maze from above, avoid all the traps, and know all the shortcuts. For noobs like me I run into all the deadends and then have to backtrack and try another route.

My strong pitch is we make the maze simpler. As with all codebases the complexity will increase over time, look at Theano. But we have this one opportunity right to make things simpler so please lets please do it, if not for you for me :heart: 

This is a bigger discussion than this PR so if you have thoughts let move it to more conversational medium.

In the meanwhile though, I really appreciate you both giving me handwritten notes of how to navigate things. Think of it likes clues for the maze that help me, and hopefully future users, navigate things as well as you two :)

**Comment by canyon289** [comment]
Updated per comments. Verified that its compiling locally. @OriolAbril tagging you per request, also to ensure you feel like the changes youd like have been made

image

**Comment by OriolAbril** [comment]
still have mixed feelings about the "named array" term, the caption looks good otherwise, but we should wait until fixing CI to check the preview on readthedocs.

I have one question though. Are those terms/definitions used elsewhere in the documentation? We could include those in the glossary and here so that when they appear somewhere else we can use `term` role and render it as a link to its glossary definition

File: docs/source/learn/examples/dimensionality.ipynb
Comment (michaelosthege) [inline docs/source/learn/examples/dimensionality.ipynb]:
```suggestion
    "+ *Coords* → A dictionary mapping dimension names to coordinate values\n",
```
Related, this is how I defined [it in the McBackend protobuf](https://github.com/michaelosthege/mcbackend/blob/2706f8570e395bdcc006111911f6716d73ecc5ee/protobufs/meta.proto#L40-L46):
```proto
// A named array that labels a dimension.
message Coordinate {
    // Name of the dimension these labels belong to.
    string name = 1;
    // Coordinate values.
    npproto.ndarray values = 2;
}
```

**Comment by canyon289** [comment]
> still have mixed feelings about the "named array" term, the caption looks good otherwise, but we should wait until fixing CI to check the preview on readthedocs.
> 
> I have one question though. Are those terms/definitions used elsewhere in the documentation? We could include those in the glossary and here so that when they appear somewhere else we can use `term` role and render it as a link to its glossary definition

Great Idea ill add them to the glossary

**Comment by canyon289** [comment]
@OriolAbril actually can we create an issue ticket to add the terms to the glossary and merge this as is? This will be a great issue for the data umbrella sprint, and worst case I'll just get back to it afterwards.

RE: Name arrays that has been changed by Michael so we should be good to go there as well

**Review by OriolAbril — APPROVED** [review]
> @OriolAbril actually can we create an issue ticket to add the terms to the glossary and merge this as is? This will be a great issue for the data umbrella sprint, and worst case I'll just get back to it afterwards.

We can take this over in a follow up PR, and if you have time please open an issue. But I don't think it's a good issue for the sprint. It's not only a matter of copying definitions from one place to another but about making sure definitions appear in both places (pulling them from a single source of truth though) as they are key to understanding the doc, and having to click a link to read them would be a disservice.

**Comment by canyon289** [comment]
Ill definitely open an issue to track.

As for this not being good for the sprint curious why would it wouldnt be? Its a straightforward change and takes folks through all the steps of compiling the docs and making a pr? Not implying you need to change your mind, just curious what I'm missing

**Comment by canyon289** [comment]
Note, I'll merge this after a rebase on main to make sure Ci is passing. Will get to it in next ~72 hours

**Comment by OriolAbril** [comment]
The key is in the _from a single source of truth_ syntagma. IMO both the glossary and this page should show the complete definitions, and then other pages using these terms should link to the glossary using the `term` role. 

The first idea that comes to mind is using myst substitutions for this which would mean writing the definitions in conf.py and then having this page and the glossary. None of the webinars cover this and there are no references on how to do this anywhere in our docs, you need to know myst. Sphinx is amazing and very flexible, but I think one should first use it with a provided configuration, and once you know how to take advantage of the available features and settings and only then look at the conf.py file. 

I think this is also the main reason the jupyterbook people decided to use a yaml file instead of trying to teach juypterbook users to use conf.py. I find the yaml too limiting and wouldn't use it (and you can do the same jupyterbook does and more -or less!- with conf.py directly) but I see the yaml as a very valuable resource to get people to use sphinx straight away without getting lost in conf.py which can be overwhelming and is also very tempting place for premature optimization.

**Comment by OriolAbril** [comment]
extra note: I don't think CI will pass until after we merge https://github.com/pymc-devs/pymc/pull/5449

**Comment by canyon289** [comment]
> The key is in the _from a single source of truth_ syntagma. IMO both the glossary and this page should show the complete definitions, and then other pages using these terms should link to the glossary using the `term` role.
> 
> The first idea that comes to mind is using myst substitutions for this which would mean writing the definitions in conf.py and then having this page and the glossary. None of the webinars cover this and there are no references on how to do this anywhere in our docs, you need to know myst. Sphinx is amazing and very flexible, but I think one should first use it with a provided configuration, and once you know how to take advantage of the available features and settings and only then look at the conf.py file.
> 
> I think this is also the main reason the jupyterbook people decided to use a yaml file instead of trying to teach juypterbook users to use conf.py. I find the yaml too limiting and wouldn't use it (and you can do the same jupyterbook does and more -or less!- with conf.py directly) but I see the yaml as a very valuable resource to get people to use sphinx straight away without getting lost in conf.py which can be overwhelming and is also very tempting place for premature optimization.

Thanks, makes sense now. 

Regarding CI if it wont pass I'll just merge this as theres zero chance broke anything in the main library

**Comment by canyon289** [comment]
Merging

**Pull request merged**
